Cookbook/Proposals/SplitOctopus.md

```diff
@@ -0,0 +1,40 @@
+# Split Octopus codebase
+
+* Author(s): Giorgos Tsiapaliokas
+* Review Managers: David Rodrigues, Yasuhiro Inami, Olivier Halligon
+
+## Introduction
+
+From the early versions of the codebase Octopus was the place were features were added.
+Unfortunately Octopus hasn't adapted as the rest of the codebase has.
+In the iOS codebase we use frameworks in order to organize the code and Octopus isn't one.
+
+## Motivation
+
+In it's current state Octopus is collection of services and features which is consumed
+by each target this requires some manual work for each target.
+
+## Proposed solution
+
+We propose to convert Octopus into a framework with the name `BabylonOctopus`.
+By having Octopus as a framework it will be easier for each target to make use of Octopus' features and services.
+With this proposal there is still the need for splitting Octopus even further which will be addressed by other proposals
+in the future.
+
+### Transition Period
+
+Converting Octopus will require some time since it's not trivial.
+We propose to  initially create an empty framework and then move each subdirectory of Octopus piece by piece.
+This way it will be easier for the team to review the changes.
```

Comment:
If we choose to only have one framework at first (and keep the splitting of that framework into smaller pieces for a later proposal as you suggest), then it would be nice to at least ensure that each piece of code moved into the new framework is independent of each other. There's no infallible way to do that and make the compiler ensure that for us if we keep a single framework though (unlike if we create multiple independent frameworks right away), but we could still mention this "keep the various features of Octopus independent from each other" as a hard goal to keep in mind here during those step-by-step migration.

Also we could mention that we could try to organise the `BabylonOctopus` framework in folders that would likely correspond to independent features so that when we decide to later split that framework into smaller ones in a future proposal, we already have pieces of code mainly grouped appropriately and ready to be split. i.e. that proposal could suggest an initial list of groups and how we could organise that framework in independent folders – even if that split isn't final and makes no guarantee and will evolve as we move the code piece by piece into that framework and realise we might need to tweak this folder hierarchy as we progress on that migration, at least that would be a good guideline to start with.

Reply:
I am not convinced that we will be able to move all subdirectories from the main target to a new framework. I don't think it will be feasible to move all UI code.

Also, GP @ Hand related code should be moved to `BabylonNHS`. It might make sense to start with this refactoring.

Reply:
Yeah that's also why I think it would be beneficial to already have a view of what will go where (in the new `BabylonOctopus/BabylonAppServices` framework and if so in which subfolder? Or in another framework like `BabylonNHS`?), that work of moving the code from the main target should be the occasion to also reorganise how the files are split in the main target today to dispatch them in what might end up being a different folder hierarchy and organisation.

My point (and I think yours too) being that since it will likely not be a 1-to-1 exact match from "folder X of main target goes into same folder X of new framework", but instead will likely be "some files of folder X of main targets goes into folder Y of new framework, some other files of folder X will go into a separate folder Z of new framework, and other files will move to BabylonNHS".

And I think we need to make the exercise and agree to such a mapping (even if it's not a definitive one, at least to have a rough idea) before approving the proposal, both to understand if such a mapping makes sense and if keeping the proposed solution (only create one framework now and split later) is the best choice or not, but also to have a sense of the amount of work that will be needed to do that migration and split, in order to vote if the proposed plan makes sense

Reply:
I wonder if it would be possible to declare a custom `module.modulemap` file that would automatically create one submodules per top-level folder we'd put in that Octopus framework.
That way we'd still only have one single `BabylonOctopus` target and one framework, but the content of each folder would be forced to be independent (and we could even require to explicitly `import BabylonOctopus.FeatureA` to only import the features/swift files that are in the `FeatureA` folder), that would offer us the guarantee that each folder would have to be independent – Xcode would tell us if we need to declare some explicit dependencies between those folders – without having to create separate frameworks right away, and with the advantage that if we see a file that needs to be moved from one folder to another while we work on this refactoring, it's just a matter of moving a file between folder (and let the compiler tell us if that would accidentally create unwanted dependencies between those submodules), without having to commit to a particular fixed and frozen folder organisation but letting us experiment with various splits as we advance on the piece-by-piece move...

Not even sure if that's totally possible – we'd need to look at blog posts and documentation about `module.modulemap` to be sure – but might be worth looking into the idea?